internal_ws/ykcompile/src/arch/x86_64/mod.rs

```diff
@@ -123,45 +123,121 @@ macro_rules! binop_add_sub {
+    //     >1 byte, then we must also save and restore RDX.
+    //   - If our register allocator has put the second SIR operand in a register that will be
+    //     clobbered, then we cannot use this register as the r/m operand. In this case we borrow
+    //     RCX temporarily.
     ($name: ident, $op:expr) => {
         fn $name(&mut self, opnd1_reg: u8, opnd2: &IRPlace) {
-            // mul and div overwrite RAX, RDX, so save them first.
+            let size = SIR.ty(&opnd2.ty()).size();
+            // For all but r/m8 variants, RDX is clobbered.
+            if size > 1 {
+                dynasm!(self.asm
+                    ; push rdx
+                );
+            }
+            // RAX is an always clobbered.
             dynasm!(self.asm
                 ; push rax
-                ; push rdx
-                ; xor rdx, rdx
             );
+            // Set up first operand.
             dynasm!(self.asm
                 ; mov rax, Rq(opnd1_reg)
             );
-            let size = SIR.ty(&opnd2.ty()).size();
+            // Set up second operand.
             let src_loc = self.iplace_to_location(opnd2);
             match src_loc {
-                Location::Reg(r) => match size {
-                    1 => {
+                Location::Reg(src_r) => {
+                    // Handle cases where the second operand was stored in an already clobbered
+                    // register. In these cases we load the value back from the stack.
+                    let (r, borrow_rcx) = if src_r == RAX.code() {
+                        let rax_off = i32::try_from(QWORD_REG_SIZE).unwrap();
                         dynasm!(self.asm
-                            ; $op Rb(r)
+                            ; push rcx
+                            ; mov rcx, [rsp + rax_off]
                         );
-                    }
-                    2 => {
+                        (RCX.code(), true)
+                    } else if size > 1 && src_r == RDX.code() {
+                        let rdx_off = i32::try_from(QWORD_REG_SIZE).unwrap() * 2;
                         dynasm!(self.asm
-                            ; $op Rw(r)
+                            ; push rcx
+                            ; mov rcx, [rsp + rdx_off]
                         );
-                    }
-                    4 => {
+                        (RCX.code(), true)
+                    } else {
+                        (src_r, false)
+                    };
+                    // Zero unused extended operand space.
+                    if size > 1 {
                         dynasm!(self.asm
-                            ; $op Rd(r)
+                            ; xor rdx, rdx
                         );
                     }
-                    8 => {
+                    match size {
+                        1 => {
+                            dynasm!(self.asm
+                                ; and ax, 0xff // Zero unused extended operand space.
```

Comment:
Didn't we do this on line 199 (or, at least, the comment matches)?

Reply:
We did it for size > 1.

Reply:
We could move it up in an else branch if you like?

Reply:
Dumb question: if we did it for size >1 on line 199, why do we then do it for size==1 here? Couldn't we handle both cases on line 199?

Reply:
We raced with the same idea.

I only sneaked the `and` in there to make the code shorter, but at the cost of readability I admit.

1fadfad fixes it.